crates/lib/src/client.rs

```diff
@@ -93,6 +102,8 @@ impl Qcs {
         let channel = get_channel(uri)?;
         let service =
             wrap_channel_with_retry(wrap_channel_with(channel, self.get_config().clone()));
+        #[cfg(feature = "grpc-web")]
+        let service = wrap_channel_with_grpc_web(service);
```

Comment:
As implemented here, if a Rust user enables the "grpc-web" feature, then their clients will all be built as `grpc-web` clients. This works transparently without any additional configuration from the user. This might be a reasonable approach if we look at the Rust crate in a vacuum, but it's flawed for `qcs-sdk-python`.

In order to expose a feature like `grpc-web` to Python, we have to build the Python package with that feature enabled for everyone. This is because most users depend on a pre-built wheel being available for their platform, which means we have to build the package up front, with a static set of features. However, this would mean that all qcs-sdk-python users would be calling our services via a grpc-web proxy, which isn't desirable. 

That leads us to modifying the feature such that it enables access to the `grpc-web` client, but still requires opt-in via code. We could expose a `use_grpc_web` on the `Qcs` client, but that doesn't quite work. Each `wrap_channel_with_*` method wraps the passed in type in another type. For example, `wrap_channel_with_grpc_web(T)` returns a `GrpcWebType<T>`. This means we can't conditionally build the client based on a `use_grpc_web` flag, because if it wasn't set, the function would return `T`, but if it was set, the function would need to return `GrpcWebType<T>`, which is a trick Rust doesn't allow.

It may be possible to work around this with something like traits or generics, but I wasn't able to fenagle a solution. `tonic` has an [issue tracking implementation of a generic Client trait](https://github.com/hyperium/tonic/issues/1490) but until that exists, I don't think we can effectively solve this problem, but I would appreciate second opinions.

Here are some potential solutions I've been kicking around, curious to hear other ideas too:

* If `grpc-web` is enabled, add a `use_grpc_web` option to the `Qcs` client and an additional method that will wrap the client with the grpc-web service. Users of the `Qcs` client would be required to check the value of the option and use the new method to wrap the client if it's enabled. This is obviously not ideal and error prone. However, all the exposed methods would be `pub(crate)`, so it's still an implementation detail we have to get right, there isn't a risk to the user to misuse the client.
* Similar to the above, but make `use_grpc_web` a request level option. That is, functions like `get_quilt_calibrations` would take a `use_grpc_web` parameter. This makes the responsibility of branching based on the flag more clear, but clutters the function signatures. I think the UX isn't as good either. If I'm a user using `grpc-web`, it's more than likely because I need it for all requests, so having to set it on every request is cumbersome.

Reply:
Is it possible to use an opaque wrapper around `Arc<Mutex<dyn GrpcService<BoxBody>>>` for the Python side? Looking at e.g. the `TranslationClient` docs, the inner client to `TranslationClient::new` is a generic `T` implementing `GrpcService<BoxBody>`. There are associated types as well, so it would really look like

```rust
// For some concrete types T, E, and F
Arc<Mutex<dyn GrpcService<BoxBody, ResponseBody=T, Error=E, Future=F>>>
```

Reply:
So the overall structure of this feature would look like:

- The api client has a feature that enables the ability to use (but does not force) grpc-web
- The SDK (Rust) offers the same
- The client-creation methods (such as above) have a parameter to select whether to use `grpc-web` (maybe we introduce a `ClientOptions` or similar)
- The type alias `GrpcConnection` becomes `Box<dyn GrpcService<BoxBody, ResponseBody=T, Error=E, Future=F>>` for some concrete types `T`, `E`, and `F`. This should work because `GrpcService` is implemented for [all types implementing `Service`](https://docs.rs/tonic/0.9.2/tonic/client/trait.GrpcService.html#impl-GrpcService%3CReqBody%3E-for-T) with certain generics, and `Service` [is implemented](https://docs.rs/tower-service/0.3.2/tower_service/trait.Service.html#impl-Service%3CRequest%3E-for-Box%3CS%3E) for `Box<S> where S: Service` 

I believe this should work, as all of our client wrappers use the same `Pin<Box<dyn Future>>` for `F` and `E` should be the same. I'm unsure if `T` will be the same, but I think it should be.

Reply:
Third option: can we just publish `qcs-sdk-python-webgrpc` as a separate PyPI package? I realize that ~doubles our Maturin build time.

Reply:
> Third option: can we just publish `qcs-sdk-python-webgrpc` as a separate PyPI package? I realize that ~doubles our Maturin build time.

Yes, I believe we could, the two concerns that come to mind are:

1. How do we specify the dependency in downstream packages like pyQuil appropriately. E.g. it needs qcs-sdk-python or qcs-sdk-python-webgrpc. I'm not sure if there is an easy way to express that.
2. That approach doesn't scale well with additional features. If we wanted to support a second feature, we'd need 4 packages total. A third feature, 6, etc.

Reply:
I guess I'm thinking of http/1.1 support less as a "feature" that makes our services better and more as a hack needed by an important client. So there'd be a separate PyPI package for `pyquil-webgrpc`, too, and we wouldn't have a matrix of feature support because this is a one-off.

That might not be the best way to look at this, though, especially since web-grpc does have the nice side-effect of giving us a way to use the API from a browser.

Reply:
@BatmanAoD using a separate `pyquil-grpc-web` ultimately became the plan 👍